Feat/comms behaviour protocol

elenaf9 opened the pull request:
# Description of change

Implementation of a custom Network Behaviour protocol for stronghold-communication.

Replace the currently used Network Behaviour with a custom protocol. The new protocol follows the basic concepts of the libp2p-request-response protocol, but additionally implements a firewall and directly integrates and leverages libp2p-relay and libp2p-mdns.

## Type of change

- [x] Enhancement (a non-breaking change which adds functionality)
- [x] Documentation Fix


## Change checklist

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

elenaf9 committed:
feat(comms): init new network behaviour protocol

Start implementation of a new Network Behaviour protocol based on the
currently used libp2p RequestResponse protocol.
Refactor the event types that are used in the protocol.

elenaf9 committed:
feat(comms): integrate mdns + relay to behaviour

Add the MDNS and relay network behaviours to the new network behaviour
protocol analogous to how the NetworkBehaviour macro implements it.

elenaf9 committed:
feat(comms): peer connection and addrs management

Add PeerConnectionManager to manage for peers the known addresses and pending requests on the current connections.

elenaf9 committed:
feat(comms): send Res from outbound handshake via channel

Return the response for an outbound request via channel instead of
emitting a Behaviour Event for it.

elenaf9 committed:
feat(comms): remove events for outbound responses

Bubble incoming requests up and emit as BehaviourEvent, with a channel
to respond. Do not emit events for ResponseSent or InboundFailures,
since there is no way to re-send a response, and it's the responsibility
of the remote to retry by re-sending the request.

elenaf9 committed:
feat(comms): rename types

Rename:
- struct MessageProtocol -> CommunicationProtocol
- struct RequestResponse -> NetBehaviour
- trait MessageEvent -> RqRsMessage
- struct RequestResponseHandler -> ConnectionHandler

elenaf9 committed:
feat(comms): add firewall, manage pending requests

Add Firewall configuration, and a mpsc channel to explicitly ask for
peer rules and approvals if necessary. Verify for each inbound and
outbound request that they are approved before forwarding them.
Add RequestManager to cache requests that are waiting for approval or a
connection to the remove peer, integrate the PeerConnectionsManager into
the RequestManager. Track all pending events with their current status
in the RequestManager, manage FIFO queue for required Network Actions.

elenaf9 committed:
chore(comms): update libp2p v0.37 -> v0.38

elenaf9 committed:
test(comms): add firewall tests

elenaf9 committed:
feat(comms): manage peer addresses and relays

Maintain a list of relay peers that are tried if a peer can not be
connected directly. Manage the list of known addresses for each peer,
their currently used endpoints (if connected), and whether a relay is
used.

elenaf9 committed:
fix(comms): no implicit fallback for using relays

If for one peer a specific relay is set, but no address for that relay
is known, it should not fallback on using other relays.
Instead, an DialError will be returned.

elenaf9 committed:
test(comms): addresses, dialing and using relays

elenaf9 committed:
feat(comms): clean code

Clean exports, format imports.
Add `RequestMessage` type for inbound/outbound Request-Response `Query`s.

elenaf9 committed:
feat(comms): add interface for the comms-library

Add `ShCommunication` as interface for the library to abstract from the
libp2p logic. Use one mpsc channel to forward incoming request, and one
for queries to the firewall.
Add a first basic test for the interface.

elenaf9 committed:
feat(comms): interface: publish network events

Optionally publish relevant SwarmEvents and Inbound / Outbound-Failures
from the interface through a mpsc channel.
Change the firewall test to use the interface.

elenaf9 committed:
fix(comms): fix no-default-features, fix interface

Fix errors when disabling the mdns feature.
Allow graceful shutdown of the initially spawned thread in the
interface.
Report also the NetworkEvents that are produced by proactive
method calls.
Make behaviour module private.

elenaf9 committed:
test(comms): test dialing via interface

elenaf9 committed:
fix(workflows): temp check comms-refactor folder

Temporarily change the Github workflows to test
/communication-refactored instead of /communication.
Will be reverted once communication-refactored replaces the old
communication.

elenaf9 committed:
feat(comms): add Errs, move types mod to interface

mxinden commented:
Great to see usage of rust-libp2p in general and libp2p-relay in specific! I would be really curious to hear any feedback on libp2p-relay (and of course on the greater rust-libp2p project). Also note the effort around [circuit relay v2 implementation](https://github.com/libp2p/rust-libp2p/pull/2059).

mxinden commented:
In case you are interested in hole punching for stronghold.rs, the following resources might be interesting:

- [Long term vision document](https://github.com/mxinden/specs/blob/hole-punching/connections/hole-punching.md)
- Status quo hole punching tracking issue across libp2p implementations https://github.com/libp2p/specs/issues/312
- Status quo hole punching tracking issue for rust-libp2p https://github.com/libp2p/rust-libp2p/issues/2052

Let me know in case you need more info. Happy to schedule a call or continue in some async fashion.

elenaf9 committed:
docs(comms): add documentation, clean code

Add comments to all relevant methods.
Emit `ProtocolSupport` to Handler from `NetBehaviour`, instead of sending
the firewall rules.
Add smaller bugfixes.

elenaf9 committed:
Merge branch 'dev' of github.com:iotaledger/stronghold.rs into feat/comms-behaviour-protocol

elenaf9 committed:
feat(comms): add Builder for ShCommunication

Add ShCommunicationBuilder for customization of the Swarm.
Add docs for interface types.

elenaf9 committed:
fix(comms): extend / fix docs, clean code

Improve docs.
Allow setting the CommunicationProtocol version, to allow support of
different versions.
Add feature flag for using the tcp-transport.
Fix test_dialing.

elenaf9 committed:
feat(comms): return Outbound Err via Response Chan

Feat: Return `OutboundFailures` as error via `response_rx` channel in the
`ResponseReceiver`, instead of emitting a `NetworkEvent::OutboundFailure`
for it. Handle the response channel of outbound requests in the `NetBehaviour`
and bubble up inbound responses from the low layer `RequestProtocol`.

Fix: assign inbound requests in the `PeerConnectionManager` to the actual
connection they were received on.

elenaf9 committed:
feat(comms): add outbound failure for shutdown

elenaf9 committed:
feat(comms): use concurrent hashmap in interface

elenaf9 committed:
fix(comms): use wasm_timer in handler

elenaf9 committed:
feat(comms): refactor interface

Refactor `ShCommunication`:
- Use tokio runtime
- All swarm interaction is handled in its own task `SwarmTask`
- `ShCommunication` is wrapper for a mpsc::Sender:
  sends `SwarmOperation` commands to swarm-task to initiate actions
- result for operations are returned via oneshot channel

Add feature flags for using relay-protocol and tcp-transport.

Temporarily use rust-libp2p v0.39 from GitHub.

elenaf9 committed:
fix(comms): fix listeners

elenaf9 committed:
feat(comms): Use Fn in FW-Rule to approve Requests

Change Firewall `Rule`:
Remove `Rule::Permission` which enforced using the `RequestPermission`
macro and related concepts. Allow approving / rejecting requests based
on a function in `Rule::Restricted.
Required `Clone` to be implemented for `Rq` - required for using
`FirewallRequest::RequestApproval` logic.

Remove outdated notes.md file.

elenaf9 committed:
Merge branch 'dev' of github.com:iotaledger/stronghold.rs into feat/comms-behaviour-protocol

elenaf9 committed:
feat(comms): clean code, firewall types

 Allow using tailored version of a request to check approval in the
 firewall.
 Remove obsolete firewall types `ToPermissionVariants`, `PeerRuleQuery`
 and `RequestApprovalQuery`.
 Do not query `FirewallRequest::PeerSpecificRule` for a specific
 direction, instead apply the returned `FirewallRules` for both
 directions.

 Improve interface docs.

 stronghold-derive: `RequestPermissions`: implement
 `Borrow<RequestPermission>` for `Request`.

 Use libp2p v0.39 from crates.io.

elenaf9 commented:
@mxinden in case that you have time (feel free to say no in case that you are busy!), we would really appreciate if we could get some feedback from you on the new interface of our communication library (`communication-refactored/src/interface.rs` / `communication-refactored/src/interface/swarm_task.rs`. 

The general purpose of the interface is to handle the polling of the swarm and provide a convenient way to perform operations on the swarm without blocking.
The interface (`ShCommunication`) is implemented by handling all Swarm interaction in a single task( `SwamTask`). Operations are triggered by sending an according message through a mpsc channel to the swarm-task, and return the outcome via an oneshot channel that is included in the message.  Inbound requests, and general swarm events (/ `NetEvent`) are forwarded through a `mpsc` channel.

It would be great to to know whether you think that our implementation is a valid way to abstract from the underlying rust-libp2p logic, or if you have any concerns with it.

mxinden commented:
> @mxinden in case that you have time (feel free to say no in case that you are busy!), we would really appreciate if we could get some feedback from you on the new interface of our communication library (communication-refactored/src/interface.rs / communication-refactored/src/interface/swarm_task.rs.

Of course. I will give this pull request a review later this week.

elenaf9 committed:
feat(comms): actix CommunicationActor draft

Add a draft for the Communication Actor on top of the `ShCommunication`
interface, using the actix framework.
Handle inbound request by implementing `StreamHandler` for the actor.
Handle async interface operations by using `ResponseActFuture` as Result
in Handler.
Get Client Add from a generic `ActorRegistry`.

Add minimal test for `CommunicationActor`.

elenaf9 committed:
feat(comms): ARegistry as Service; add impl_handler macro

Expect Actor-Registry to be ArbiterService, get its Addr by using the
`from_registry` method.
Use macro to implement Handler for different Request types.

mxinden commented on communication-refactored/src/actor.rs:
```diff
@@ -0,0 +1,166 @@
+// Copyright 2020-2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+#![cfg(feature = "actor")]
+
+pub mod messages;
```
```suggestion
#![cfg(feature = "actor")]
pub mod messages;

```

The `cfg` only applies to the single line, not the whole module, right?

elenaf9 replied:
It was meant for the whole module. 
But I can actually remove that completely, since I already have in my lib.rs:
```
#[cfg(feature = "actor")]
pub mod actor;
```

mxinden commented on communication-refactored/src/behaviour.rs:
```diff
@@ -0,0 +1,1024 @@
+// Copyright 2020-2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+// Copyright 2020 Parity Technologies (UK) Ltd.
```
:pray:

mxinden commented on communication-refactored/src/behaviour.rs:
```diff
@@ -0,0 +1,1024 @@
+            self.handle_updated_peer_rule(peer, RuleDirection::Both);
+        }
+
+        // Handle individual approvals fro requests that were returned after a [`FirewallRequest::RequestApproval`]
```
```suggestion
        // Handle individual approvals for requests that were returned after a [`FirewallRequest::RequestApproval`]
```

mxinden commented on communication-refactored/src/interface.rs:
```diff
@@ -0,0 +1,559 @@
+/// The local [`PeerId`] is derived from public key of the IdKeys.
+/// If this is not the case, remote Peers will reject the communication.
+pub enum InitKeypair {
+    /// Identity Keys that are be used to derive the noise keypair and peer id.
```
```suggestion
    /// Identity Keys that are used to derive the noise keypair and peer id.
```

mxinden commented on communication-refactored/src/interface.rs:
```diff
@@ -0,0 +1,559 @@
+        if let Some(limit) = self.connections_limit {
+            swarm_builder = swarm_builder.connection_limits(limit);
+        }
+        let swarm = swarm_builder.build();
```
```suggestion
        let swarm = swarm_builder.substream_upgrade_protocol_override(Version::V1Lazy).build();
```

Consider using `V1Lazy` instead of `V1` to save on a whole network roundtrip during stream negotiation, e.g. when doing a request.

See https://docs.rs/multistream-select/0.10.2/multistream_select/enum.Version.html#variant.V1Lazy for details.

elenaf9 replied:
Thanks for the advice. I do have to admit though that the docs on `V1Lazy` are a bit difficult to understand:
> This strategy is only applicable [...] if the dialer supports just a single application protocol

What does `a single application protocol` mean? That only one version of a protocol is supported?

> Nevertheless, there is a pitfall that is rarely encountered: When nesting multiple protocol negotiations, the listener should either be known to support all of the dialer’s optimistically chosen protocols or there is must be no intermediate protocol without a payload and none of the protocol payloads must have the potential for being mistaken for a multistream-select protocol message

In what cases are protocol negotiations nested? And are there any protocols in rust-libp2p without a payload?

mxinden replied:
> > This strategy is only applicable [...] if the dialer supports just a single application protocol
> 
> What does `a single application protocol` mean? That only one version of a protocol is supported?

Protocols on streams are negotiated using [Multistream Select](https://github.com/multiformats/multistream-select). The process is quite simple. The dialer proposes a protocol, the listener either accepts the protocol by echoing it, or rejects it by sending `na` (not available).

Say the dialer opens a new stream and only has a single protocol that it is willing to speak on that stream. E.g. not `/my-protocol/1` and `/my-protocol/2`, but only a single one, e.g. `/my-protocol/2`. What the dialer can do is send the `/my-protocol/2` and instead of waiting for the echo, or the `na`, directly send application data, not having to wait for 1 round trip.

This only works, when one is willing to speak a single protocol on that stream only. When willing to speak either one of many protocols, `V1Lazy` defaults to the `V1` behavior.


> 
> > Nevertheless, there is a pitfall that is rarely encountered: When nesting multiple protocol negotiations, the listener should either be known to support all of the dialer’s optimistically chosen protocols or there is must be no intermediate protocol without a payload and none of the protocol payloads must have the potential for being mistaken for a multistream-select protocol message
> 
> In what cases are protocol negotiations nested? And are there any protocols in rust-libp2p without a payload?

Negotiations are nested e.g. when first negotiating a security protocol, then, nested, a multiplexer and then, nested a stream protocol. `substream_upgrade_protocol_override(Version::V1Lazy)` only sets the `V1Lazy` option for the last one, not followed by any other nesting, thus safe.

I am not aware of any **intermediate** protocols without a payload.

---

This is just a suggestion. After all, playing it safe by using `V1` is just fine. You can still later on enable `V1Lazy` without having to coordinate a roll out.

elenaf9 replied:
Thank you for the detailed explanation, that makes sense. Using `V1Lazy` sounds reasonable.

elenaf9 replied:
For some reason, `test test_firewall::firewall_ask` does not terminate when using `V1Lazy`, and if disable the feature `tcp-transport` (which results in the use of a plain `TokioTcpConfig` as transport), `test_firewall::firewall_permissions` does not terminate as well. All other tests (dialing, sending requests) pass.

Since I don't really know what the reason for this behaviour could be, I'll use `V1` for now. 
I might investigate this further in the future, but right now it is not a priority.

mxinden replied:
>  I might investigate this further in the future, but right now it is not a priority.

In case you get around to it, I would be curious what the cause is.

mxinden commented on communication-refactored/src/interface.rs:
```diff
@@ -0,0 +1,559 @@
+    command_tx: mpsc::Sender<SwarmOperation<Rq, Rs, TRq>>,
+}
+
+impl<Rq, Rs, TRq> ShCommunication<Rq, Rs, TRq>
```
This looks like a great interface for others to integrate with.

mxinden commented on communication-refactored/src/interface/swarm_task.rs:
```diff
@@ -0,0 +1,571 @@
+                    request,
+                    response_tx,
+                };
+                let _ = self.request_tx.try_send(received_rq);
```
Is sending these to the _inbound request handler_ on a best-effort basis good enough? Why not properly send them via `SinkExt::send`?

elenaf9 replied:
But wouldn't `SinktExt::send` require to `await` the returned future, hence potentially block everything? 
And I am not sure how this would behave in case that the `Receiver` is not dropped, but also never read. 

I have considered adding a RingBuffer structure to cache requests in case that `try_send` returns `SendErrorKind::Full`.  In that case the loop in `SwarmTask::run` would also always do something like
```
futures::select! {
      ...
      _ = poll_fn(|cx| request_tx.poll_ready(cx)).fuse() => {
          // send oldest element from ring buffer
      }
  }
```
The benefit from using a ring buffer would be that only the most recent requests would be cached.
Another option would be to use an unbounded channel for inbound requests, but this could potentially use a lot of memory and even be a security risk if a remote is spam-sending requests.

I so far kept it rather simple because usually I'd expect the `Receiver` to be in some sort of loop that continuously reads from it (and if the `Receiver` is dropped there nothing we can do anyway). But I agree that this is probably not good enough.
What do you think about the ring-buffer idea @mxinden?

mxinden replied:
> But wouldn't SinktExt::send require to await the returned future, hence potentially block everything?

That is correct. You would depend on the other side to continuously read from the channel. Is there reason to worry that the other side will not continuously read from that channel?

> Another option would be to use an unbounded channel for inbound requests, but this could potentially use a lot of memory and even be a security risk if a remote is spam-sending requests.

I would strongly advice against unbounded channels. See for example https://github.com/paritytech/substrate/search?q=unbounded&type=issues for the many issues we had related to unbounded channels.

>  What do you think about the ring-buffer idea @mxinden?

While I think it is a healthy attitude to drop old messages in favor of new ones, I don't quite see the need for the ring buffer in the first place, circling back to my question above, is there reason to worry that the other side will not continuously read from that channel?

elenaf9 replied:
> Is there reason to worry that the other side will not continuously read from that channel?

In general, I would expect the receiver to continuously read from the channel, yes. 
But I could also imagine use-cases in which a request may require e.g. some asynchronous operations or user interaction, and where requests are handled sequentially. In this case, a next request might only be read from the channel after the preceding one was resolved, which could then potentially take some time. In the meantime, no swarm interaction could take place once the capacity of the channel was reached. I think this behaviour may be desired in some cases, but maybe not in every one.

And if the receiver continuously polls the request channel, `SinktExt::send` would not block, but in this case the current best-effort basis via `Sender::try_send` would also be sufficient, wouldn't it? 

--- 
I guess my general motive is that the `SwarmTask` should not have to rely on any behaviour outside of it's own loop. Because it is running in its own decoupled task, I'd like to guarantee that it will not block, which is why I try to not use any asynchronous operations (apart from the central `run` method).

mxinden replied:
_Take the below with a grain of salt. After all I am not very familiar with your whole stack._

> I think this behaviour may be desired in some cases, 

It would be desired to enforce back-pressure throughout the system. E.g. when the machine is already at its limit, the network should pause for a bit and not produce even more events.

> but maybe not in every one.

I would expect a large enough channel size to cater for those cases.

> And if the receiver continuously polls the request channel, SinktExt::send would not block, but in this case the current best-effort basis via Sender::try_send would also be sufficient, wouldn't it?

It would. Though (a) you loose the first mentioned back-pressure (whether that is useful in your case or not I can't tell) and (b) silently dropped events are hard to debug.

elenaf9 replied:
> It would be desired to enforce back-pressure throughout the system. E.g. when the machine is already at its limit, the network should pause for a bit and not produce even more events.

That's a good point, which I did not consider before.

---
After thinking about this more, I decided to go the extra mile and support all 3 discussed configuration (block/drop/use ring-buffer). I think this behaviour should be decided by the user, especially since Stronghold is only a library and I do not know all use cases.
I have added a new commit https://github.com/iotaledger/stronghold.rs/pull/210/commits/d93c9f632bbcba336ae70e2b18b86e0e286b5111 that implements a wrapper for the `mpsc::Sender` and implements `Sink` for it according to the set config.

mxinden reviewed (commented):
I haven't found the time yet to give this a full review. Submitting an intermediate one for now. I will take another look in the upcoming days.

I like the _Interface_ and _Swarm Task_ abstraction.

elenaf9 commented:
Thanks @mxinden!

mxinden commented on communication-refactored/src/interface.rs:
```diff
@@ -0,0 +1,559 @@
+
+    /// Start listening on the network.
+    /// If no address is given, the listening address will be OS-assigned.
+    pub async fn start_listening(&mut self, address: Option<Multiaddr>) -> Result<Multiaddr, ListenErr> {
```
When e.g. listening on `/ip4/0.0.0.0/tcp/0` multiple listen addresses might be returned, no?

elenaf9 replied:
Yes that's true. The problem is that I don't see a way of determining when the last "new-listen-address" is reported / when I have "collected" all addresses associated with a new listener. I am currently only returning the address from the first reported `SwarmEvent::NewListenAddr` of the associated Listener.
Do you see a way of solving this?

But either way, I will correct my docs, since it currently implies that there is only one listening address, which is not correct in cases like the one you mentioned.

mxinden replied:
Given that a listener might emit a `NewAddress` at any point in time, the only way I see to properly report this is a channel instead of a oneshot, that lives as long as the listener, being used for each `NewListenAddr`.

That said, I am not sure it is worth the complexity in your use-case.

elenaf9 committed:
feat(comms): Executor config, fix start_listening

Allow configuration of a custom executor to use in libp2p::Swarm and
for spawning the `SwarmTask`.

Change `start_listening` to always require a Multiaddress, instead of
optionally using an OS-assigned one. This is necessary because a
transport may be used where `/ip4/0.0.0.0/tcp/0` is not be supported.

Adjust tests to the changes.

mxinden commented:
One note on the usage of Cargo features to enable/disable mdns and relay: Features are additive, thus in case this crate is imported twice with different feature sets, the union of the two will be used.

elenaf9 committed:
refactor(p2p): rename library communication -> p2p

Rename:
- Package `stronghold-communication` -> `stronghold-p2p`
- Interface `ShCommunication` -> `StrongholdP2p`
- Actor `CommunicationActor` -> `NetworkActor`
- `CommunicationProtocol` -> `MessageProtocol`

elenaf9 committed:
feat(p2p): wrap  channel for inbound requests/ events

Add `EventChannel`, which wraps the `mpsc::Receiver` of the channels that
are used by the `SwarmTask` to forward inbound requests and network
events.
Support different `ChannelSinkConfig`s in case that the channel is full
when a new event occurs:
- block until there is capacity
- drop new event
- buffer event, drop older buffered events if favor of new ones.

mxinden commented on p2p/src/interface/msg_channel.rs:
```diff
@@ -0,0 +1,180 @@
+                        buffer.pop_front();
+                    }
+                    buffer.push_back(e.into_inner());
+                    if let Some(waker) = self.waker.as_ref() {
```
```suggestion
                    if let Some(waker) = self.waker.take() {
```

I am not 100% sure, though I would expect to be woken up at most once when waiting for a new item to be added to the `buffer`, no?

elenaf9 replied:
I actually had it like this at first, but because I was also not 100% sure I changed it to not consume the waker. 

But you are right. If a new item was added to the buffer and the task was waken again, it will attempt to send the request. Thus wakening the task is then related to the channel being ready, and not to new items in the buffer.

elenaf9 committed:
feat(p2p): integrate Network-Actor into Stronghold

Move NetworkActor from stronghold-p2p crate into iota-stronghold crate.
Add missing functionality to Network-Actor.
Integrate NetworkActor into Stronghold interface, to allow read/ write/
executing procedure on a remote Stronghold.

elenaf9 committed:
fix(comms): remove outdated stronghold-comms crate

felsweg-iota commented on client/Cargo.toml:
```diff
@@ -21,8 +21,11 @@ zeroize = "1.1"
+futures = { version = "0.3", optional = true }
+actix = "0.12"
+
+# remove dependency
```
let's better remove this commented out dependency to riker.

felsweg-iota commented on client/src/interface.rs:
```diff
@@ -486,93 +577,87 @@ impl Stronghold {
+    /// Spawn the p2p-network actor and swarm.
+    pub async fn spawn_p2p(&mut self, firewall_rule: Rule<ShRequest>, network_config: NetworkConfig) -> StatusMessage {
+        if self.network_actor.is_some() {
+            return StatusMessage::Error(String::from("Network actor was already spawned"));
```
We can think of further extending the error types, or even having a `P2PError`. Would be good to handle certain errors explicitly.

elenaf9 replied:
I agree, but at the moment it is returning a String for the Error in order to be consistent with the return values of the other interface methods. 
I think we should do a separate PR, that changes the general interface to return `anyhow::Error`s in the `ResulMessage::Error`, and add a P2PError-type in that PR. 
Thoughts @felsweg-iota ?

elenaf9 replied:
Ah I just noticed that with the recent changes in the interface, it already returns an anyhow Error.
I will add the P2PError then and bubble up the actual error types of the p2p library.

felsweg-iota replied:
I agree on specific error types, where they make sense to provide a more consistent way of handling them.

felsweg-iota commented on client/src/interface.rs:
```diff
@@ -486,93 +577,87 @@ impl Stronghold {
-            )
-            .expect(line_error!());
-        self.communication_actor = Some(communication_actor);
+        let network_actor = unwrap_or_err!(NetworkActor::new(self.target.clone(), firewall_rule, network_config).await);
```
Let's think on moving the convenience macros to crate level.

felsweg-iota commented on client/src/interface.rs:
```diff
@@ -486,93 +577,87 @@ impl Stronghold {
 
-#[cfg(feature = "communication")]
+#[macro_export]
+macro_rules! unwrap_or_err (
```
Let's think of using error types rather than strings.

elenaf9 replied:
See comment above.

felsweg-iota commented on client/src/interface.rs:
```diff
@@ -585,324 +670,227 @@ impl Stronghold {
+            request: CallProcedure { proc: control_request },
+        };
+        match actor.send(send_request).await {
+            Ok(Ok(Ok(res))) => res,
```
hmm is there a way to unwrap it in a more clear way? This nesting of results seem redundant, maybe we can work around that?

felsweg-iota commented on p2p/src/behaviour.rs:
```diff
@@ -0,0 +1,1031 @@
+        self.addresses.remove_relay(peer);
+    }
+
+    #[cfg(feature = "relay")]
```
we can think of moving all feature gates parts of the code into an own module.

felsweg-iota commented on p2p/src/behaviour.rs:
```diff
@@ -0,0 +1,1031 @@
+
+    // Query for individual approval of a requests.
+    // This is necessary if the firewall is configured with [`Rule::Ask`].
+    fn query_rq_approval(&mut self, peer: PeerId, request_id: RequestId, rq: TRq, direction: RequestDirection) {
```
Better use words eg. `query_request_approval`

felsweg-iota commented on p2p/src/behaviour.rs:
```diff
@@ -0,0 +1,1031 @@
+}
+
+#[cfg(test)]
+mod test {
```
Better move tests outside the module.

elenaf9 replied:
This is not possible here, as it is testing the `NetBehaviour`, which is not exposed to the outside of the crate.

felsweg-iota commented on p2p/src/behaviour/firewall/permissions.rs:
```diff
@@ -0,0 +1,94 @@
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct PermissionValue(u32);
+
+impl PermissionValue {
```
This could be generic, that `PermissionValue` would have generic Type `T`, that is being comparable and hashable. This way we can keep the number of permissions dynamic. I can't see a reason why the permissions have a purely numeric value that is being stored inside an unsigned int. Does the inherent ordering have a semantic value inside the firewall rules?

elenaf9 replied:
In general, the logic inside of this `permissions.rs` module is not used anymore in the stronghold-p2p library itself. Instead it provides some logic for convenience, that may be used by the user when writing a `Rule::Restricted. An example for this implemented in the test `test_firewall::firewall_permissions`.

But it being a numeric value is on purpose yes. The idea is (similar to how the `rwx` permissions in unix systems implement it) that each variant has a numeric value to the power of 2, e.g. ReadStore: 1, WriteStore: 2, WriteVault: 4, ExecuteProcedure: 8, ..., so that the a `FirewallPermission` can be created from the sum of allowed `PermissionValue`s.
It could be implemented that the type of integer is generic, but I decided against it because 1. It's unlikely that there are request messages enum with more than 32 variants , and 2. the user can always do their own implementation if this doesn't fit their use-case.

felsweg-iota replied:
I do see the point, and would leave it. In general I would advise to always be open towards more options and flexibility.

felsweg-iota commented on p2p/src/behaviour/request_manager.rs:
```diff
@@ -0,0 +1,573 @@
+    marker::PhantomData,
+};
+
+// Actions for the behaviour to handle i.g. the behaviour emits the appropriate `NetworkBehaviourAction`.
```
```suggestion
// Actions for the behaviour to handle eg. the behaviour emits the appropriate `NetworkBehaviourAction`.
```

felsweg-iota reviewed (commented):
That was a big PR and a good one. Good work! From my side the code is good so far.

elenaf9 committed:
Merge branch 'dev' of github.com:iotaledger/stronghold.rs into HEAD

elenaf9 committed:
fix(p2p): fix documentation

lucas-tortora reviewed (approved):
LGTM!

elenaf9 committed:
fix(covector): remove comms library from covector

elenaf9 committed:
fix(p2p): clean code

felsweg-iota commented on client/src/interface.rs:
```diff
@@ -843,20 +820,17 @@ impl Stronghold {
-            Ok(Ok(Err(e))) => ProcResult::Error(e.to_string()),
-            Ok(Err(e)) => ProcResult::Error(e.to_string()),
-            Err(e) => ProcResult::Error(e.to_string()),
+        let receive_response = unwrap_or_err!(actor.send(send_request).await);
```
nice!

felsweg-iota reviewed (approved):
good one.
